docs/src/main/sphinx/admin/fault-tolerant-execution.rst

```diff
@@ -0,0 +1,344 @@
+========================
+Fault-tolerant execution
+========================
+
+By default, if a Trino node lacks the resources to execute on a task or
+otherwise fails during query execution, the query fails and must be run again
+manually. The longer the runtime of a query, the more likely it is to be
+susceptible to such failures.
+
+Fault-tolerant execution is a mechanism in Trino that enables a cluster to
+mitigate query failures by retrying queries or their component tasks in
+the event of failure. With fault-tolerant execution enabled, in-progress query
+data is spooled and can be re-used by another worker in the event of a worker
+outage or other fault during query execution.
+
+.. note::
+
+    Fault tolerance does not apply to broken queries or other user error. For
+    example, Trino does not spend resources retrying a query that fails because
+    its SQL cannot be parsed.
+
+Configuration
+-------------
+
+Fault-tolerant execution is disabled by default. To enable the feature, set the
+``retry-policy`` configuration property to either ``QUERY`` or ``TASK``
+depending on the desired :ref:`retry policy <fte-retry-policy>`.
+
+.. code-block:: properties
+
+    retry-policy=QUERY
+
+The following configuration properties control the behavior of fault-tolerant
+execution on a Trino cluster:
+
+.. list-table:: Fault-tolerant execution configuration properties
+   :widths: 30, 40, 30
+   :header-rows: 1
+
+   * - Property name
+     - Description
+     - Default value
+   * - ``retry-policy``
+     - Configures what is retried in the event of failure, either
+       ``QUERY`` to retry the whole query, or ``TASK`` to retry tasks
+       individually if they fail. See :ref:`retry policy <fte-retry-policy>` for
+       more information.
+     - ``NONE``
+   * - ``exchange.deduplication-buffer-size``
+     - Size of the buffer for data sent from worker nodes to the coordinator. If
```

Comment:
As I understand it's a buffer in memory on the coordinator that's used to store exchange data that it can re-send to a worker node in the event of failure, to continue the query. So by that logic I believe it's a grand total for any one query.

Would appreciate clarification on that if I'm wrong though @losipiuk @linzebing

Reply:
The buffer is used to store outputs of the coordinator stages. In two cases 32MB is not enough:
1. A SELECT query whose result size is larger than 32MB
2. An INSERT query that writes to a lot of files/partitions that the metadata of them exceeds 32MB

So actually 32MB is sufficient in most cases.